### DISCUSSION TITLE
WindowNative.GetWindowHandle no longer returns valid hWnd

### DISCUSSION THREAD

**Discussion by BertrandDeSaintBezier** [unoplatform/uno#18775] [General]
We are trying to embed a Win32 window in a Uno application as a User Control. In previous versions of Uno (5.1.80, for example, see attached screenshot) this worked fine, where the following method : [`WinRT.Interop.WindowNative.GetWindowHandle(object)`](https://learn.microsoft.com/en-us/windows/apps/develop/ui-input/retrieve-hwnd) would return a valid `Window Handle Id`, which could be attached to the Uno process using [`nint SetParent(nint child, nint parent)`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setparent).

**Uno WinUI : 5.1.80** 
![image](https://github.com/user-attachments/assets/9d92e46e-b28a-4fe1-809e-4977cbe156bc)

It seems like in the newer Uno SDK, this same method doesn't seem to return a coherent Window Handle (see attached screenshot in version 5.5.32), and we are unable to hook the Win32 process to Uno. It appears that the implementation of `GetWindowHandle` has changed to : `return new IntPtr((long)window.AppWindow.Id.Value);`

**Uno.Sdk: 5.5.32**
![image](https://github.com/user-attachments/assets/d3d8c2c3-ea15-4bd1-b080-dfebc643d698)

We found [this ongoing discussion on the WinAppSDK repo](https://github.com/microsoft/WindowsAppSDK/discussions/3463) and figured that maybe we could use the `Win32Interop.GetWindowFromWindowId(appWindow.Id)` method to get the Id, but we don't have access to the Win32Interop class.

Not sure if this is a recent Uno implementation change or something to do with WinAppSDK, but we have found a workaround using Reflection, where we are brute-forcing our way to the window's hWnd Id.

```csharp
Window? mainWindow = (Microsoft.UI.Xaml.Application.Current as App)!.MainWindow;
object nativeWindow = mainWindow.GetType().GetProperty("NativeWindow", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(mainWindow);
object hwndSource = nativeWindow.GetType().GetProperty("HwndSourceWindow", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(nativeWindow);

applicationHwnd = (IntPtr)hwndSource.GetType().GetProperty("Handle", BindingFlags.Instance | BindingFlags.Public).GetValue(hwndSource);
```

Thought we would mention this in case it was a bug, and to help anyone else that has been running into a similar issue.

Cheers :)

**Comment by jeromelaban** [comment]
Thanks for the question.

To be sure, which target framework are you debugging with?

**Reply by BertrandDeSaintBezier** [reply]
Hi Jérome, sorry for the late answer. This was compiled for Windows/Desktop

**Reply by jeromelaban** [reply]
Thanks, but on which target framework specifically? `netX.0-desktop` or `netX.0-windows10.YYY`? The latter we won't be able to fix if there was a change.

/cc @MartinZikmund

**Reply by BertrandDeSaintBezier** [reply]
In our case, the issue is on the `net8.0-desktop` framework, as we are on the latest version of Uno SDK. 

Just for completedness' sake, I can confirm that this is working perfectly fine in older versions (the one I use is `net8.0-windows10.0.19041`), so the issue only appears to be with recent versions.

**Reply by MartinZikmund** [reply]
The problem here is that in case of `-desktop` the API is using our own implementation, which allows us to use this method against internal windows, instead of native ones. This can be adjusted though. On the other hand, the `windows.10.XYZ` version should still work as you intended.

**Reply by BertrandDeSaintBezier** [reply]
As we are currently migrating from the older `net8.0-windows10.xxx` versions to the newer `net8.0-desktop` ones, what would be your recommendation to get this working ? 

Going back to the original issue, we can't use the Id returned to us by `WindowNative.GetWindowhandle()` (in other words, the _Interal Window_ Id) to parent a Win32 window to it, which is why we are doing it using reflection (though we would prefer not to 😅). Are there any plans to add a new method which would return the _Native Window_ Id ?

**Reply by MartinZikmund** [reply]
I created https://github.com/unoplatform/uno/issues/18842 to track this, but not sure how quickly I will be able to get to it. The important point is that `net8.0-windows10.xxx` is not an "older" target, it is actually just a different target altogether - it targets Windows using WinUI 3 directly (e.g. no Uno Platform is involved there, everything runs against Microsoft's framework directly); vs `-desktop` which is our own cross-platform desktop target framework. So if you are targeting Windows specifically, then the `-windows` target is still a good choice

**Reply by BertrandDeSaintBezier** [reply]
Okay sounds good. Thanks for clarifying this point ! I will close the discussion for now and keep tabs on the Issue 😄
